### PR TITLE
chore: Use bindings.v1File instead of ContextItem

### PR DESCRIPTION
**Pull request by rb-determined-ai**
This is an early part of a larger refactor.  Overall, the goal is to
move away from handcrafted API requests towards using the generated
bindings.

One blocker to that effort is that the creation of context directories
is definitely based on the handcrafted APIs.

Additionally, the context loading code has grown "crusty" over time; it
involved a lot of bandaid-style coversions between old and new formats,
in addition to the fact that a Context class existed, but it didn't
really do anything.  This PR focuses on migrating to a bindings-native
storage structure and reducing overall "crustiness".

### CODE REVIEW

File: master/internal/core_experiment.go [hunk 1]
```diff
@@ -378,14 +378,10 @@ func (m *Master) parseCreateExperiment(params *CreateExperimentParams, user *mod
...
-	// CreateExperimentParams has highest priority (coming from WebUI)
+	// CreateExperimentParams has highest priority.
 	projectID := 1
 	if params.ProjectID == nil {
-		if config.Workspace() == "" && config.Project() != "" {
-			return nil, false, nil,
-				errors.New("workspace and project must both be included in config if one is provided")
-		}
-		if config.Workspace() != "" && config.Project() == "" {
+		if (config.Workspace() == "") != (config.Project() == "") {
```
Comment (rb-determined-ai) [inline master/internal/core_experiment.go:384]: logically equivalent, reduces code duplication

File: harness/determined/common/context.py [hunk 2]
```diff
@@ -1,223 +1,173 @@
...
+        gid=0,
+    )
+
+
+def read_v1_context(
+    local_path: pathlib.Path,
+    limit: int = constants.MAX_CONTEXT_SIZE,
+) -> List[bindings.v1File]:
     """
+    Retrun a list of v1Files suitable for submitting a context directory over the v1 REST API.
```
Comment (mpkouznetsov) [inline harness/determined/common/context.py:73]:
```suggestion
    Return a list of v1Files suitable for submitting a context directory over the v1 REST API.
```

File: harness/determined/common/context.py [hunk 3]
```diff
@@ -1,223 +1,173 @@
...
+        items.append(f)
+        size += v1File_size(f)
+
+    local_path = local_path.resolve()
+
+    if not local_path.exists():
+        raise Exception("Path '{}' doesn't exist".format(local_path))
+
+    if local_path.is_file():
+        raise ValueError("Path '{}' must be a directory".format(local_path))
```
Comment (mpkouznetsov) [inline harness/determined/common/context.py:94]: f-string?

File: harness/determined/common/context.py [hunk 4]
```diff
@@ -1,223 +1,173 @@
...
+    root_path = local_path
+
+    ignore = list(constants.DEFAULT_DETIGNORE)
+    ignore_path = root_path.joinpath(".detignore")
+    if ignore_path.is_file():
+        with ignore_path.open("r") as detignore_file:
+            ignore.extend(detignore_file)
+    ignore_spec = pathspec.PathSpec.from_lines(pathspec.patterns.GitWildMatchPattern, ignore)
+
+    msg = "Preparing files (in {}) to send to master... {} and {} files".format(
```
Comment (mpkouznetsov) [inline harness/determined/common/context.py:105]: f-string?

File: harness/determined/common/experimental/determined.py [hunk 5]
```diff
@@ -78,34 +78,38 @@ def create_experiment(
...
-        model_context, _ = context.read_context(model_dir)
+        model_context = context.read_v1_context(model_dir)
 
-        resp = self._session.post(
-            "/api/v1/experiments",
-            json={
-                "config": yaml.safe_dump(experiment_config),
-                "model_definition": model_context,
-            },
+        req = bindings.v1CreateExperimentRequest(
```
Comment (mpkouznetsov) [inline harness/determined/common/experimental/determined.py:99]: 👍

File: master/internal/core_experiment.go (hunk 1 above)
Comment (mpkouznetsov) [reply master/internal/core_experiment.go:384]: 👍

**Review by mpkouznetsov — APPROVED** [review]
a few nitpicks but basically great!

File: harness/determined/common/context.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/common/context.py:94]: sure.  I just indented this code and didn't see it, but I'm happy to fix them.

**Pull request merged**